docs/source/en/chat_templating.md

```diff
@@ -298,10 +298,100 @@ one? If that happens, you should:
+as it is one of the highest-performing tool-use models in its size category at the time of writing. If you have the
+memory, you can consider using a larger model instead, like [Command-R](https://huggingface.co/CohereForAI/c4ai-command-r-v01)
+or [Mixtral-8x22B](https://huggingface.co/mistralai/Mixtral-8x22B-Instruct-v0.1), both of which also support tool use.
 
+First, let's load our model and tokenizer:
+
+```python
+import torch
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+checkpoint = "NousResearch/Hermes-2-Pro-Llama-3-8B"
+
+tokenizer = AutoTokenizer.from_pretrained(checkpoint, revision="pr/13")
+model = AutoModelForCausalLM.from_pretrained(checkpoint, torch_dtype=torch.bfloat16, device_map="auto")
+```
+
+Next, let's define a list of tools. For simplicity, we'll just have a single tool in this example:
+
+```python
+def get_current_temperature(location: str, unit: str) -> float:
+    """
+    Get the current temperature at a location.
+    
+    Args:
+        location: The location to get the temperature for, in the format "City, Country"
+        unit: The unit to return the temperature in. (choices: ["celsius", "fahrenheit"])
+    Returns:
+        The current temperature in the specified units, as a float.
+    """
+    return 22.  # Your real function should probably actually get the temperature!
+
+tools = [get_current_temperature]
+```
+
+Now, let's set up a conversation for our bot:
+
+```python
+messages = [
+  {"role": "system", "content": "You are a bot that responds to temperature queries. You should choose the unit used in the queried location."},
+  {"role": "user", "content": "Hey, what's the temperature in Paris right now?"}
+]
+```
+
+Now, let's apply the chat template and generate a response:
+
+```python
+inputs = tokenizer.apply_chat_template(messages, chat_template="tool_use", tools=tools, add_generation_prompt=True, return_dict=True, return_tensors="pt")
+inputs = {k: v.to(model.device) for k, v in inputs.items()}
+out = model.generate(**inputs, max_new_tokens=128)
+print(tokenizer.decode(out[0][len(inputs["input_ids"][0]):]))
+```
+
+And we get:
+
+```text
+<tool_call>
+{"arguments": {"location": "Paris, France", "unit": "celsius"}, "name": "get_current_temperature"}
+</tool_call><|im_end|>
+```
+
+The model has called the function with valid arguments, in the format requested by the function docstring. It has
+inferred that we're most likely referring to the Paris in France, and it remembered that, as the home of SI units,
+the temperature in France should certainly be displayed in Celsius.
+
+Let's append the model's tool call to the conversation, followed by the result of calling the tool. Remember, in 
+reality this is the point where you'd actually call the function, rather than just using a dummy
+result!
+
+```python
+messages.append({"role": "assistant", "tool_calls": [{"name": "get_current_temperature", "arguments": {"location": "Paris, France", "unit": "celsius"}}]})
```

Comment:
`tool_calls` is missing a level here and `arguments` is actually a string, addionally you should include a call ID (generated by you), it should be something like this:
```json
[
    {
        "id": "call__0_get_current_temperature_cmpl-...",
        "type": "function",
        "function": {
            "name": "get_current_temperature",
            "arguments": "{\"location\": \"Paris, France\", \"unit\": \"celsius\"}"
        }
    }
]
```

The call ID can then be used in the `tool` message below to allow the model to match up call and response:
```json
{
    "role": "tool",
    "content": "22.0",
    "tool_call_id": "call__0_get_current_temperature_cmpl-..."
}
```

Note that I didn't include `name` as that is not a part of the `tool` response (it was part of the deprecated `function` response) as `tool_call_id` will identify the exact call, however I guess it doesn't hurt including it if some models require it.

Reply:
Not really opposed to allowing `arguments` as a dict either, I'm guessing OpenAI did this so they wouldn't have to parse the JSON output from the model when they were using `function` and it just transferred over, but you kinda have to parse the JSON for `tool` so it's less hassle for the user to be able to just pass the dict.

Reply:
Yes, Mistral does this, check their inference example (or my chat template implemention of it):
https://github.com/mistralai/mistral-common/blob/b995dc8301987f7302f6cd1675fad30e38ce4d73/examples/tokenizer.ipynb?short_path=ce509b9#L133-L140

The model may issue a number of parallel tool calls, f.ex. if you ask `What's the weather like in Oslo and Stockholm?`, so this is a crucial feature, and Mistral handles this perfectly from my testing.

Reply:
This is how I render tool messages in my Mistral chat template:
```jinja
{% elif message['role'] == 'tool' %}
    {{ '[TOOL_RESULTS] ' }}
    {{ dict(call_id=message['tool_call_id'], content=message['content'])|tojson }}
    {{ '[/TOOL_RESULTS]' }}
{% elif message['role'] == 'assistant' %}
    {% if message['tool_calls'] %}
        {{ '[TOOL_CALLS] [' }}
        {% for call in message['tool_calls'] %}
            {% if call['type'] == 'function' %}
                {{ dict(id=call['id'], name=call['function']['name'], arguments=call['function']['arguments'])|tojson }}
            {% endif %}
            {% if not loop.last %}
                {{ ', ' }}
            {% endif %}
        {% endfor %}
        {{ ']' }}
    {% else %}
        {{ message['content'] }}
    {% endif %}
    {{ eos_token }}
{% endif %}
```

Reply:
Noted! The Mixtral-8x22B chat template in HF didn't have this, but I obviously need to investigate their actual library. Working on it!

Reply:
@CISC it's updated! I haven't updated the model chat templates to match, but what do you think of the example in the docs now?

Reply:
Looking at the docs
```python
tool_call_id = "vAHdf3"  # Random ID, should be unique for each tool call
tool_call = {"name": "get_current_temperature", "arguments": {"location": "Paris, France", "unit": "celsius"}}
messages.append({"role": "assistant", "tool_calls": [{"id": tool_call_id, "type": "function", "function": tool_call}]})
```
shouldn't `arguments` be a string? i.e.
```python
tool_call = {"name": "get_current_temperature", "arguments": '{"location": "Paris, France", "unit": "celsius"}'}
```

Reply:
This was discussed above - we're probably going to keep it as-is!